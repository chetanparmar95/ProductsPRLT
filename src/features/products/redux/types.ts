
export const FETCH_PRODUCTS = 'FETCH_PRODUCTS';
export const FETCH_PRODUCTS_SUCCESS = 'FETCH_PRODUCTS_SUCCESS';
export const FETCH_PRODUCTS_FAILED = 'FETCH_PRODUCTS_FAILED';

export const ADD_TO_CART = 'ADD_TO_CART';
export const REMOVE_FROM_CART = 'REMOVE_FROM_CART';
export const INCREMENT_QUANTITY = 'INCREMENT_QUANTITY';
export const DECREMENT_QUANTITY = 'DECREMENT_QUANTITY';