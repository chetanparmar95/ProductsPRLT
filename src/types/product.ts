export interface Product {
    id: number;
    colour: string;
    name: string;
    price: number;
    img: string;
    quantity: number;
}