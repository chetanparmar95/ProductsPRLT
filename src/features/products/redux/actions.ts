import { Product } from "../../../types/product";
import { ADD_TO_CART, DECREMENT_QUANTITY, FETCH_PRODUCTS, FETCH_PRODUCTS_FAILED, FETCH_PRODUCTS_SUCCESS, INCREMENT_QUANTITY, REMOVE_FROM_CART } from "./types";

export const fetchProducts = () => {
    return { type: FETCH_PRODUCTS }
}

export const fetchProductsSuccess = (payload: Product[]) => {
    const products = payload.map(product => {
        return {...product, quantity: 0}
    })
    return { type: FETCH_PRODUCTS_SUCCESS, payload: products}
}

export const fetchProductsFailed = (error: any) => {
    return { type: FETCH_PRODUCTS_FAILED, payload: error }
}

export const addItemToCart = (product: Product) => {
    return { type: ADD_TO_CART, payload: product}
}

export const removeItemFromCart = (productId: number) => {
    return { type: REMOVE_FROM_CART, payload: {id: productId}}
}

export const incrementQuantity = (productId: number) => {
    return { type: INCREMENT_QUANTITY, payload: {id: productId}}
}

export const decrementQuantity = (productId: number) => {
    return { type: DECREMENT_QUANTITY, payload: {id: productId}}
}