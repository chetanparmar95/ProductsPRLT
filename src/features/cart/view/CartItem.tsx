import React from 'react';
import { Image, StyleSheet, Text, TouchableOpacity, View } from "react-native"
import { Product } from '../../../types/product';
import { useDispatch } from 'react-redux';
import { decrementQuantity, incrementQuantity, removeItemFromCart } from '../../products/redux/actions';
import Button from '../../../components/button';

interface Props {
    item: Product
}

const CartItem = ({item}: Props) => {
    const dispatch = useDispatch();
    return (
        <View key={item.id} style={styles.cartItem}>
            <Image source={{ uri: item.img }} style={styles.cartItemImage} resizeMode='contain' />
            <View style={styles.cartItemDetails}>
                <Text style={styles.cartItemName}>{item.name}</Text>
                <Text style={styles.cartItemPrice}>Price: ${item.price.toFixed(2)}</Text>
                <View style={styles.bottomRow}>
                    <View style={styles.quantityControl}>
                        <TouchableOpacity
                            testID='decrement'
                            style={styles.quantityButton}
                            onPress={() => item.quantity && item.quantity > 1 ? dispatch(decrementQuantity(item.id)) : dispatch(removeItemFromCart(item.id))}>
                            <Text style={styles.quantityButtonText}>-</Text>
                        </TouchableOpacity>
                        <Text style={styles.quantityText}>{item.quantity}</Text>
                        <TouchableOpacity
                            testID='increment'
                            style={styles.quantityButton}
                            onPress={() => dispatch(incrementQuantity(item.id))}>
                            <Text style={styles.quantityButtonText}>+</Text>
                        </TouchableOpacity>
                    </View>
                    <Button title='Remove' containerStyle={styles.removeButtonContainer} onPress={() => dispatch(removeItemFromCart(item.id))} />
                </View>
            </View>
        </View>
    )
}

export default CartItem

const styles = StyleSheet.create({
    cartItem: {
        flexDirection: 'row',
        alignItems: 'center',
        marginBottom: 20,
    },
    cartItemImage: {
        width: 80,
        height: 120,
        borderRadius: 10,
    },
    cartItemDetails: {
        flex: 1,
        marginLeft: 15,
    },
    cartItemName: {
        fontSize: 18,
        fontWeight: 'bold',
    },
    cartItemPrice: {
        fontSize: 16,
        color: '#888',
        marginTop: 5,
    },
    bottomRow: {
        flexDirection: 'row', 
        justifyContent: 'space-between', 
        marginTop: 12, 
        alignItems: 'center'
    },
    quantityControl: {
        flexDirection: 'row',
        alignItems: 'center',
    },
    quantityButton: {
        backgroundColor: '#FF5722',
        borderRadius: 15,
        width: 30,
        height: 30,
        justifyContent: 'center',
        alignItems: 'center',
    },
    quantityButtonText: {
        color: '#fff',
        fontSize: 20,
    },
    quantityText: {
        fontSize: 18,
        marginHorizontal: 10,
    },
    removeButtonContainer: {
        paddingVertical: 0,
        height: 30
    }
});