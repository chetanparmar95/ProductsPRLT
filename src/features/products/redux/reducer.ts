import { Reducer } from "redux";
import { Product } from "../../../types/product";
import { ADD_TO_CART, DECREMENT_QUANTITY, FETCH_PRODUCTS_FAILED, FETCH_PRODUCTS_SUCCESS, INCREMENT_QUANTITY, REMOVE_FROM_CART } from "./types";

export interface ProductReducerState {
    products: Product[],
    cartItems: Product[],
    error: any
}
const initialState: ProductReducerState = {
    products: [],
    cartItems: [],
    error: null
};

const productReducer: Reducer<ProductReducerState> = (state = initialState, action: any) => {
    switch (action.type) {
        case FETCH_PRODUCTS_SUCCESS:
            const products = action.payload?.map((p: Product) => ({...p, quantity: 0}))
            return {
                ...state,
                products: products,
            };
        case FETCH_PRODUCTS_FAILED:
            return {
                ...state,
                error: action.payload,
            };
        case ADD_TO_CART:
            const exist = state.cartItems.find(item => item.id === action.payload.id)
            if(!exist) {
                const product = {...action.payload, quantity: 1}
                return {
                    ...state,
                    cartItems: [...state.cartItems, product]
                }
            }
            return { ...state }
        case REMOVE_FROM_CART:
            const filteredList = state.cartItems.filter(item => item.id !== action.payload.id)
            return {
                ...state,
                cartItems: filteredList
            }
        case INCREMENT_QUANTITY:
            const { id: incrementId } = action.payload;
            const incrementedCartItems = state.cartItems.map((item) => {
                if (item.id === incrementId) {
                    return { ...item, quantity: item.quantity + 1 };
                }
                return item;
            });
            return { ...state, cartItems: incrementedCartItems };
        case DECREMENT_QUANTITY:
            const { id: decrementId } = action.payload;
            const decrementedCartItems = state.cartItems.map((item) => {
                if (item.id === decrementId && item.quantity && item.quantity > 1) {
                    return { ...item, quantity: item.quantity - 1 };
                }
                return item;
            });
            return { ...state, cartItems: decrementedCartItems };
        default:
            return state;
    }
};

export default productReducer;