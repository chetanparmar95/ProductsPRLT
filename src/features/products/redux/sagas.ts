import { call, put, takeLatest } from 'redux-saga/effects'
import ProductServices from '../../../services/products';
import { fetchProductsFailed, fetchProductsSuccess } from './actions';
import { FETCH_PRODUCTS } from './types';

// exporting for UT purpose
export function* fetchProducts(): Generator<any> {
   try {
      const response: any = yield call(ProductServices.getProducts);
      yield put(fetchProductsSuccess(response.data));
   } catch (e: any) {
      yield put(fetchProductsFailed(e));
   }
}

function* productSaga() {
  yield takeLatest(FETCH_PRODUCTS, fetchProducts);
}

export default productSaga;