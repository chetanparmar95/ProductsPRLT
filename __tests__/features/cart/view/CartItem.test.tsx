
import React from 'react';
import { fireEvent, render } from '@testing-library/react-native';
import { Provider } from 'react-redux';
import configureStore from 'redux-mock-store';
import CartItem from '../../../../src/features/cart/view/CartItem';
import { Product } from '../../../../src/types/product';
import { decrementQuantity, incrementQuantity, removeItemFromCart } from '../../../../src/features/products/redux/actions';

const mockStore = configureStore([]);

describe('CartItem Component', () => {
  it('CartItem renders correctly', () => {
    const item: Product = {
      id: 1,
      name: 'Product 1',
      colour: 'black',
      price: 10.0,
      img: 'https://example.com/product1.jpg',
      quantity: 2,
    };
    const store = mockStore({});
    const { toJSON } = render(
      <Provider store={store}>
        <CartItem item={item} />
      </Provider>
    );
    expect(toJSON()).toMatchSnapshot();
  });

  it('CartItem renders values correctly', () => {
    const item: Product = {
      id: 1,
      name: 'Product 1',
      colour: 'black',
      price: 10.0,
      img: 'https://example.com/product1.jpg',
      quantity: 2,
    };
    const store = mockStore({});
    const { getByText, getByTestId } = render(
      <Provider store={store}>
        <CartItem item={item} />
      </Provider>
    );

    const productName = getByText('Product 1');
    const productPrice = getByText('Price: $10.00');
    const quantityText = getByText('2');
    const decrementButton = getByTestId('decrement');
    const incrementButton = getByTestId('increment');
    
    expect(productName).toBeDefined();
    expect(productPrice).toBeDefined();
    expect(quantityText).toBeDefined();
    expect(decrementButton).toBeDefined();
    expect(incrementButton).toBeDefined();
  });

  it('should dispatch an action when increment button is pressed', () => {
    const item: Product = {
      id: 1,
      name: 'Product 1',
      price: 10.0,
      colour: 'black',
      img: 'https://example.com/product1.jpg',
      quantity: 2,
    };

    const store = mockStore({});

    const { getByText } = render(
      <Provider store={store}>
        <CartItem item={item} />
      </Provider>
    );

    // Find and click the increment button
    const incrementButton = getByText('+');
    fireEvent.press(incrementButton);

    // Check if the expected action was dispatched
    const actions = store.getActions();
    expect(actions).toContainEqual(incrementQuantity(item.id));
  });

  it('should dispatch an action when decrement button is pressed', () => {
    const item: Product = {
      id: 1,
      name: 'Product 1',
      price: 10.0,
      colour: 'black',
      img: 'https://example.com/product1.jpg',
      quantity: 2,
    };

    const store = mockStore({});

    const { getByText } = render(
      <Provider store={store}>
        <CartItem item={item} />
      </Provider>
    );

    // Find and click the decrement button
    const decrementButton = getByText('-');
    fireEvent.press(decrementButton);

    // Check if the expected action was dispatched
    const actions = store.getActions();
    expect(actions).toContainEqual(decrementQuantity(item.id));
  });

  it('should dispatch an action when decrement button is pressed', () => {
    const item: Product = {
      id: 1,
      name: 'Product 1',
      price: 10.0,
      colour: 'black',
      img: 'https://example.com/product1.jpg',
      quantity: 1,
    };

    const store = mockStore({});

    const { getByText } = render(
      <Provider store={store}>
        <CartItem item={item} />
      </Provider>
    );

    const removeButton = getByText('Remove');
    fireEvent.press(removeButton);

    // Find and click the decrement button
    const decrementButton = getByText('-');
    fireEvent.press(decrementButton);

    // Check if the expected action was dispatched
    const actions = store.getActions();
    expect(actions).toContainEqual(removeItemFromCart(item.id));

  });
})
