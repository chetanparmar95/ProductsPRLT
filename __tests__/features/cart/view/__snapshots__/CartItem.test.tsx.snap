// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CartItem Component CartItem renders correctly 1`] = `
<View
  style={
    {
      "alignItems": "center",
      "flexDirection": "row",
      "marginBottom": 20,
    }
  }
>
  <Image
    resizeMode="contain"
    source={
      {
        "uri": "https://example.com/product1.jpg",
      }
    }
    style={
      {
        "borderRadius": 10,
        "height": 80,
        "width": 80,
      }
    }
  />
  <View
    style={
      {
        "flex": 1,
        "marginLeft": 15,
      }
    }
  >
    <Text
      style={
        {
          "fontSize": 18,
          "fontWeight": "bold",
        }
      }
    >
      Product 1
    </Text>
    <Text
      style={
        {
          "color": "#888",
          "fontSize": 16,
          "marginTop": 5,
        }
      }
    >
      Price: $
      10.00
    </Text>
    <View
      style={
        {
          "alignItems": "center",
          "flexDirection": "row",
          "justifyContent": "space-between",
          "marginTop": 12,
        }
      }
    >
      <View
        style={
          {
            "alignItems": "center",
            "flexDirection": "row",
          }
        }
      >
        <View
          accessibilityState={
            {
              "busy": undefined,
              "checked": undefined,
              "disabled": undefined,
              "expanded": undefined,
              "selected": undefined,
            }
          }
          accessibilityValue={
            {
              "max": undefined,
              "min": undefined,
              "now": undefined,
              "text": undefined,
            }
          }
          accessible={true}
          collapsable={false}
          focusable={true}
          onClick={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            {
              "alignItems": "center",
              "backgroundColor": "#FF5722",
              "borderRadius": 15,
              "height": 30,
              "justifyContent": "center",
              "opacity": 1,
              "width": 30,
            }
          }
          testID="decrement"
        >
          <Text
            style={
              {
                "color": "#fff",
                "fontSize": 20,
              }
            }
          >
            -
          </Text>
        </View>
        <Text
          style={
            {
              "fontSize": 18,
              "marginHorizontal": 10,
            }
          }
        >
          2
        </Text>
        <View
          accessibilityState={
            {
              "busy": undefined,
              "checked": undefined,
              "disabled": undefined,
              "expanded": undefined,
              "selected": undefined,
            }
          }
          accessibilityValue={
            {
              "max": undefined,
              "min": undefined,
              "now": undefined,
              "text": undefined,
            }
          }
          accessible={true}
          collapsable={false}
          focusable={true}
          onClick={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            {
              "alignItems": "center",
              "backgroundColor": "#FF5722",
              "borderRadius": 15,
              "height": 30,
              "justifyContent": "center",
              "opacity": 1,
              "width": 30,
            }
          }
          testID="increment"
        >
          <Text
            style={
              {
                "color": "#fff",
                "fontSize": 20,
              }
            }
          >
            +
          </Text>
        </View>
      </View>
      <View
        accessibilityState={
          {
            "busy": undefined,
            "checked": undefined,
            "disabled": undefined,
            "expanded": undefined,
            "selected": undefined,
          }
        }
        accessibilityValue={
          {
            "max": undefined,
            "min": undefined,
            "now": undefined,
            "text": undefined,
          }
        }
        accessible={true}
        collapsable={false}
        focusable={true}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          {
            "alignItems": "center",
            "backgroundColor": "#FF5722",
            "borderRadius": 20,
            "flexDirection": "row",
            "height": 30,
            "justifyContent": "center",
            "opacity": 1,
            "paddingHorizontal": 10,
            "paddingVertical": 0,
          }
        }
      >
        <Text
          style={
            [
              {
                "color": "#fff",
                "fontWeight": "bold",
                "marginLeft": 5,
              },
              {},
            ]
          }
        >
          Remove
        </Text>
      </View>
    </View>
  </View>
</View>
`;
